sys/include/net/gcoap/dns.h

```diff
@@ -0,0 +1,269 @@
+     * - -ENOMSG, if CoAP response did not contain a DNS response.
+     * - -ETIMEDOUT, if CoAP request timed out.
+     */
+    int res;
+    uint8_t method;         /**< CoAP method to be used for the DNS message
+                             *   exchange */
+    uint8_t dns_buf_len;    /**< Length of gcoap_dns_ctx_t::dns_buf */
+    int8_t family;          /**< Address family to resolve */
+    /**
+     * @brief The current block number for block-wise transfer
+     *
+     * Leave unset on function call.
+     */
+    uint8_t cur_blk_num;
+};
+
+/**
+ * @brief   Query a domain name via CoAPS synchronously
+ *
+ * @note    Only available with module `gcoap_dns_async` compiled in.
+ *
+ * @param[in] domain_name   A '\0'-terminated domain name. Must not be NULL.
+ * @param[out] addr_out     The resolved address. Must not be NULL.
+ * @param[in] family        The desired address family for @p addr_out.
+ *                          @ref AF_UNSPEC for any address family (an IPv6
+ *                          address will take preference over an IPv4 address).
+ * @param[in] method        CoAP method to use. Allowed values are
+ *                          `COAP_METHOD_GET`, `COAP_METHOD_FETCH`, and
+ *                          `COAP_METHOD_POST`.
+ *
+ * @return  length of @p addr_out in bytes on success
+ * @return  -EAFNOSUPPORT, if the hostname of the URI template resolves to an
+ *          unknown address family.
+ * @return  -EBADMSG, if @p method is not supported.
+ * @return  -EBADMSG, when receiving erroneous response or response containing
+ *          an error code.
+ * @return  -ECONNABORTED, if CoAP request cannot be sent.
+ * @return  -ECONNREFUSED, if no URI template is set for the client (see
+ *          @ref gcoap_dns_server_uri_template_set()).
+ * @return  -EDESTADDRREQ, if CoAP response was received from an unexpected
+ *          remote.
+ * @return  -EHOSTUNREACH, if the hostname of the URI template can not be
+ *          resolved
+ * @return  -EINVAL, if the URI template was not processable.
+ * @return  -ENOBUFS, if there was not enough buffer space for the request.
+ * @return  -ENOBUFS, if length of received CoAP body is greater than
+ *          @ref CONFIG_DNS_MSG_LEN.
+ * @return  -ENOENT, if Zone-ID of the URI template can not be found locally.
+ * @return  -ENOMSG, if CoAP response did not contain a DNS response.
+ * @return  -ENOTRECOVERABLE, on gCoAP-internal error.
+ * @return  -ENOTSUP, if credential can not be added for to client.
+ * @return  -ENOTSUP, if module `gcoap_dns_async` is not compiled in.
+ * @return  -ETIMEDOUT, if CoAP request timed out.
+ */
+int gcoap_dns_query(const char *domain_name, void *addr_out, int family,
+                    uint8_t method);
+
+/**
+ * @brief   Query a domain name via CoAPS asynchronously
+ *
+ * @param[in] domain_name   A '\0'-terminated domain name. Must not be NULL.
+ * @param[in] ctx           Context object. All members except gcoap_dns_ctx_t::res and
```

Comment:
I think that usability would be better if this would take all arguments of the synchronous form plus ctx, and allow the ctx to be fully uninitialized. This would also mean that the internals of the ctx could remain as private as can be (making future changes easier).